compiler/modules/nback/build.bal

```diff
@@ -1106,48 +1163,67 @@ final readonly & map<bir:OrderOp> flippedOrderOps = {
 };
 
 function buildCompareTagged(llvm:Builder builder, Scaffold scaffold, bir:CompareInsn insn, llvm:Value lhs, llvm:Value rhs, bir:Register result) {
-    llvm:Value lhsIsNil = builder.iCmp("eq", lhs, llvm:constNull(llvm:pointerType("i8", 1)));
-    llvm:Value rhsIsNil = builder.iCmp("eq", rhs, llvm:constNull(llvm:pointerType("i8", 1)));
-    llvm:Value eitherNil = builder.iBitwise("or", lhsIsNil, rhsIsNil);
-
-    llvm:BasicBlock eitherNilBB = scaffold.addBasicBlock();
-    llvm:BasicBlock neitherNilBB = scaffold.addBasicBlock();
-    llvm:BasicBlock joinBB = scaffold.addBasicBlock();
-    builder.condBr(eitherNil, eitherNilBB, neitherNilBB);
-
-    builder.positionAtEnd(eitherNilBB);
-    llvm:Value bothNil = builder.iBitwise("and", lhsIsNil, rhsIsNil);
+    bir:OrderType orderTy = insn.orderType;
+    llvm:Value? compareResult = ();
+    if orderTy is bir:OptOrderType {
+        match orderTy.opt {
+            t:UT_INT => {
+                compareResult = builder.call(buildRuntimeFunctionDecl(scaffold, intCompareFunction), [lhs, rhs]);
+            }
+            t:UT_FLOAT => {
+                compareResult = builder.call(buildRuntimeFunctionDecl(scaffold, floatCompareFunction), [lhs, rhs]);
+            }
+            t:UT_BOOLEAN => {
+                compareResult = builder.call(buildRuntimeFunctionDecl(scaffold, intCompareFunction), [lhs, rhs]);
+            }
+            t:UT_STRING => {
+                compareResult = builder.call(buildRuntimeFunctionDecl(scaffold, stringCompareFunction), [lhs, rhs]);
+            }
+        }
 
-    if insn.op is "<=" || insn.op is ">=" {
-        buildStoreBoolean(builder, scaffold, bothNil, insn.result);
-    }
-    else {
-        buildStoreBoolean(builder, scaffold, llvm:constInt(LLVM_BOOLEAN, 0), insn.result);
     }
-    builder.br(joinBB);
-
-    builder.positionAtEnd(neitherNilBB);
-    bir:OptOrderType orderTy = <bir:OptOrderType> insn.orderType;
-    match orderTy.opt {
-        t:UT_INT => {
-            llvm:Value lhsUntagged = buildUntagInt(builder, scaffold, <llvm:PointerValue>lhs);
-            llvm:Value rhsUntagged = buildUntagInt(builder, scaffold, <llvm:PointerValue>rhs);
-            buildCompareInt(builder, scaffold, buildIntCompareOp(insn.op), lhsUntagged, rhsUntagged, result);
+    else if orderTy is bir:ArrayOrderType {
+        match orderTy[0].opt {
+            t:UT_INT => {
+                compareResult = builder.call(buildRuntimeFunctionDecl(scaffold, arrayIntCompareFunction), [lhs, rhs]);
```

Comment:
Similar point to above.  You can have a table with an entry for each ordered simple type e.g. {UT_INT, intCompareFunction, arrayCompareFunction}.

Reply:
fixed